✨ [bento][amp-accordion] Prevent overwrite of existing ids on content and header and add a11y attributes

krdwan opened the pull request:
Bento Accordion Tracker: https://github.com/ampproject/amphtml/issues/30445

This PR does the following
1. Adds functionality so that an `id` on an existing `header` or `content` section (in either `bento` or `amp`) will not be overwritten by the components internal logic for assigning `ids`
2. Ensures `id` on header is matched with `aria-labelledby` on `content` which fixes https://github.com/ampproject/amphtml/issues/31137
3. Ensures `role` and `aria-labelledby` attributes are on the `content` and `id` is on the `header`
4. Default element tag for `header` updated from `header` to `div`.  This addresses an a11y issue in `storybook` which does not allow for `role="button"` on the `header` tag.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -316,10 +321,13 @@ export function AccordionSection({
+        headerId,
         expanded,
         expandHandler,
+        registerContentId: setContentIdState,
```
Nit: "register" sounds to me like there are many. "set" is more for one. So I think it'd be very reasonable to call this `setContentId` and `setHeaderId`.

krdwan replied:
Sounds good, updated in the latest revision.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.type.js:
```diff
@@ -56,23 +57,31 @@ AccordionDef.AccordionHeaderProps;
+ *   id: (string),
+ *   role: (string),
  *   onClick: (function()|undefined),
+ *   aria-controls: (string),
```
These two are not public API though, right? They are only set privately, so I think we don't need them here. We do not expect a developer to do `<AccordionHeader aria-expanded={...}/>`.

krdwan replied:
Per real time conversation, refactored names and added comments for non-public API here: https://github.com/ampproject/amphtml/pull/31136#issuecomment-728285011

dvoytenko commented on extensions/amp-accordion/1.0/accordion.type.js:
```diff
@@ -56,23 +57,31 @@ AccordionDef.AccordionHeaderProps;
  * @typedef {{
+ *   id: (string),
+ *   role: (string),
+ *   aria-labelledby: (string),
```
Ditto: private and thus not needed here.

krdwan replied:
Per real time conversation, refactored names and added comment for non public API here: https://github.com/ampproject/amphtml/pull/31136#issuecomment-728285011

dvoytenko commented on extensions/amp-accordion/1.0/accordion.type.js:
```diff
@@ -56,23 +57,31 @@ AccordionDef.AccordionHeaderProps;
 /**
  * @typedef {{
- *   role: (string|undefined),
+ *   id: (string),
```
`id` and `role` here and for `ContentProps` allow `undefined`

krdwan replied:
Right, let's chat about this in real time.  I did do some re-factoring of the type definitions in the restructuring PR https://github.com/ampproject/amphtml/pull/31083.  So 

- HeaderProps
- ContentProps

refer to the Header and ContentShims passed from amp side and for these, many of the props will never be undefined and the entire set of props are not public

- AccordionSectionProps
- AccordionHeaderProps
- AccordionContentProps

Refer to the actually public Accordion Components defined in `accordion.js`

krdwan replied:
Refactored names in latest revision.  Summary here: https://github.com/ampproject/amphtml/pull/31136#issuecomment-728285011

dvoytenko reviewed (changes requested):
Looks very good. Just a few comments.

caroqliu commented on extensions/amp-accordion/1.0/amp-accordion.js:
```diff
@@ -123,8 +123,14 @@ function getState(element, mu) {
-    const contentProps = dict({'as': contentShim});
+    const headerProps = dict({
+      'as': headerShim,
+      'id': section.firstElementChild.getAttribute('id'),
```
We need to add `id` as an `attributeFilter` when the mutation observer is initialized.

krdwan replied:
Added in latest revision.

krdwan commented:
Last revision includes some refactoring of names in the `.type.js` file for ease of reading:

1. AccordionContextProps -> AccordionContext
2. SectionContextProps -> SectionContext
3. HeaderProps -> HeaderShimProps (also added a comment that this is not a public API)
4. ContentProps -> ContentShimProps (also added a comment that this is not a public API)

Changes Here:
https://github.com/ampproject/amphtml/pull/31136/commits/220f865775e597c3540d7749153c47ad24b54799

dvoytenko commented on extensions/amp-accordion/1.0/accordion.type.js:
```diff
@@ -56,28 +57,40 @@ AccordionDef.AccordionHeaderProps;
  */
-AccordionDef.HeaderProps;
+AccordionDef.HeaderShimProps;
+// AccordionDef.HeaderShimProps is not a public API
```
Please merge this text with the jsdoc comment that defines the type itself. E.g.

```
/**
 * This is not a public API, it exists for ........
 * @typedef {...}
 */
```

krdwan replied:
Thanks, updated in the latest revision!

dvoytenko reviewed (approved):
LGTM with one ask
